File: examples/order-workflow.ts
/**
 * Example durable function — order processing workflow.
 *
 * Run: npx durable-viz examples/order-workflow.ts --open
 */

import {
  withDurableExecution,
  defaultSerdes,
  type DurableContext,
} from '@aws/durable-execution-sdk-js'

interface OrderEvent {
  orderId: string
  items: string[]
  total: number
  requireApproval?: boolean
}

interface ApprovalResult {
  approved: boolean
  notes?: string
}

export const handler = withDurableExecution(async (
  event: OrderEvent,
  context: DurableContext
): Promise<Record<string, unknown>> => {

  // Validate and enrich the order
  const order = await context.step('validate-order', async () => {
    return { ...event, validated: true, timestamp: new Date().toISOString() }
  })

  // Check inventory and reserve payment in parallel
  const results = await context.parallel('prepare-order', [
    {
      name: 'check-inventory',
      func: async (ctx) => {
        return await ctx.invoke('check-inventory', 'InventoryFunction', {
          items: order.items,
        })
      },
    },
    {
      name: 'reserve-payment',
      func: async (ctx) => {
        return await ctx.invoke('reserve-payment', 'PaymentFunction', {
          total: order.total,
        })
      },
    },
  ])

  // Synthesize preparation results
  const preparation = await context.step('review-results', async () => {
    return { ready: true, results: results.succeeded() }
  })

  // High-value orders need manager approval
  if (event.total > 5000) {
    const approval = await context.waitForCallback<ApprovalResult>(
      'manager-approval',
      async (callbackId) => {
        context.logger.info('Awaiting approval', { callbackId, orderId: event.orderId })
      },
      { timeout: { hours: 24 }, serdes: defaultSerdes }
    )

    if (!approval.approved) {
      return { status: 'rejected', orderId: event.orderId, notes: approval.notes }
    }
  }

  // Fulfill the order
  const shipment = await context.step('fulfill-order', async () => {
    return { shipped: true, trackingId: `TRK-${event.orderId}` }
  })

  return { status: 'completed', orderId: event.orderId, shipment }
})

File: examples/OrderWorkflow.cs
/**
 * Example durable function — order processing workflow (C#).
 *
 * Requires the preview NuGet package:
 *   dotnet add package Amazon.Lambda.DurableExecution --version 0.1.1-preview
 *
 * Run: npx durable-viz examples/OrderWorkflow.cs --open
 */

using Amazon.Lambda.Core;
using Amazon.Lambda.DurableExecution;
using Amazon.Lambda.RuntimeSupport;
using Amazon.Lambda.Serialization.SystemTextJson;

namespace OrderProcessor;

public class Function
{
    public static async Task Main(string[] args)
    {
        var handler = new Function();
        var serializer = new DefaultLambdaJsonSerializer();
        using var handlerWrapper = HandlerWrapper.GetHandlerWrapper<DurableExecutionInvocationInput, DurableExecutionInvocationOutput>(handler.Handler, serializer);
        using var bootstrap = new LambdaBootstrap(handlerWrapper);
        await bootstrap.RunAsync();
    }

    public Task<DurableExecutionInvocationOutput> Handler(
        DurableExecutionInvocationInput input, ILambdaContext context)
        => DurableFunction.WrapAsync<OrderEvent, OrderResult>(Workflow, input, context);

    private async Task<OrderResult> Workflow(OrderEvent order, IDurableContext ctx)
    {
        // Validate and enrich the order
        var validated = await ctx.StepAsync(
            async (_, _) =>
            {
                await Task.CompletedTask;
                return new ValidatedOrder(order.OrderId, order.Items, order.Total, true, DateTime.UtcNow);
            },
            name: "validate-order");

        // Reserve inventory in parallel
        var batch = await ctx.ParallelAsync(
            new[]
            {
                new DurableBranch<object>("check-inventory", async (_, _) =>
                {
                    await Task.CompletedTask;
                    return new { status = "reserved", order.Items };
                }),
                new DurableBranch<object>("reserve-payment", async (_, _) =>
                {
                    await Task.CompletedTask;
                    return new { status = "authorized", order.Total };
                }),
            },
            name: "prepare-order");

        // Wait for warehouse processing
        await ctx.WaitAsync(
            TimeSpan.FromHours(2),
            name: "warehouse-processing");

        // High-value orders need manager approval
        if (order.Total > 5000)
        {
            var approval = await ctx.WaitForCallbackAsync<ApprovalResult>(
                submitter: async (callbackId, cbCtx, _) =>
                {
                    ctx.Logger.LogInformation("Awaiting approval for {OrderId}", order.OrderId);
                    await Task.CompletedTask;
                },
                name: "manager-approval");

            if (!approval.Approved)
            {
                return new OrderResult { Status = "rejected", OrderId = order.OrderId, Notes = approval.Notes };
            }
        }

        // Fulfill the order
        var shipment = await ctx.StepAsync(
            async (_, _) =>
            {
                await Task.CompletedTask;
                return $"TRK-{order.OrderId}";
            },
            name: "fulfill-order");

        return new OrderResult { Status = "completed", OrderId = order.OrderId, TrackingId = shipment };
    }
}

public class OrderEvent
{
    public string? OrderId { get; set; }
    public string[]? Items { get; set; }
    public decimal Total { get; set; }
}

public class OrderResult
{
    public string? Status { get; set; }
    public string? OrderId { get; set; }
    public string? TrackingId { get; set; }
    public string? Notes { get; set; }
}

public record ValidatedOrder(string OrderId, string[] Items, decimal Total, bool Confirmed, DateTime Timestamp);

public class ApprovalResult
{
    public bool Approved { get; set; }
    public string? Notes { get; set; }
}
